### PR TITLE
Add eigen utils to convert mesh 3d vertices to oriented box

### PR DESCRIPTION
**Pull request by AmrElsersy**
Signed-off-by: AmrElsersy <amrelsersay@gmail.com>

# 🎉 New feature

## Summary
Adding helper functions to convert 3d vertices of a mesh to an oriented bounding box to be used in the bounding box camera 
[rendering#334](https://github.com/ignitionrobotics/ign-rendering/pull/334)

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

### CODE REVIEW

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:69]: Can you also add reference for the algorithm you used? I believe this is the link you mentioned in the past, right? http://codextechnicanum.blogspot.com/2015/04/find-minimum-oriented-bounding-box-of.html

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:80]:
```suggestion
        Eigen::Vector3d centroid = math::eigen3::convert(mean);
```
Same as my other comment about using the namespace to avoid ambiguity

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:83]:
```suggestion
        Eigen::Matrix3d covariance = computeCovarianceMatrix(_mesh, centroid);
```
The comment here is redundant since the method and variable name explain what is going on here

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:100]:
```suggestion
        projectionTransform.block<3, 3>(0, 0) = eigenVectorsPCA.transpose();
        projectionTransform.block<3, 1>(0, 3) =
```

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:101]:
```suggestion
          -1.0f * (projectionTransform.block<3, 3>(0, 0) * centroid);
```

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:107]:
```suggestion
        for (const auto &point : _mesh)
```
Pass by reference to avoid copying a `math::Vector3d` object

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:116]:
```suggestion
        const Eigen::Vector3d meanDiagonal = 0.5f * (maxPoint + minPoint);
```

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:125]: What's the significance of this comment? It's confusing me :thinking: :grimacing:

Comment (AmrElsersy) [reply eigen3/include/ignition/math/eigen3/Util.hh:125]: I was debugging something, will remove it :D

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (AmrElsersy) [reply eigen3/include/ignition/math/eigen3/Util.hh:78]: it gives me error 
```
ignition/ign-math/include/ignition/math/Helpers.hh:505:15: error: conversion from ‘int’ to non-scalar type ‘ignition::math::v6::Vector3<double>’ requested
  505 |       T sum = 0;
```

it seems that it is not suitable for double stuff

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [reply eigen3/include/ignition/math/eigen3/Util.hh:69]: +1

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:68]: need Doxygen for parameters and return value

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [reply eigen3/include/ignition/math/eigen3/Util.hh:38]: +1

**Review by scpeters — COMMENTED** [review]
you've resolved a lot of comments but I don't see the corresponding changes in the code

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (AmrElsersy) [reply eigen3/include/ignition/math/eigen3/Util.hh:65]: It was the implemntation of [PCL](https://github.com/PointCloudLibrary/pcl/blob/master/common/include/pcl/common/impl/centroid.hpp#L180)

I replaced it with the [implementation](https://github.com/isl-org/Open3D/blob/76c2baf9debd460900f056a9b51e9a80de9c0e64/cpp/open3d/utility/Eigen.cpp#L305) of `Open3D`  and it gave me the same result, let's keep it

**Comment by AmrElsersy** [comment]
> you've resolved a lot of comments but I don't see the corresponding changes in the code

forgot to push  :D

**Review by adlarkin — COMMENTED** [review]
Overall, this looks good. I'm trying to do a few verifications for testing, but I think that this can be merged as long as I don't find anything strange in the tests I run.

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:45]: nit: I would name this `_vertices` instead of `_vertex`

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:80]: nit: `_vertices`

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:79]: nit: consider rephrasing

`\brief Get the oriented 3d bounding box of a group of vertices using PCA`

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:82]: nit: `verticesToOrientedBox`

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:88]: should we check the vector size and exit early if the vector is empty to avoid dividing by zero?

**Review by scpeters — COMMENTED** [review]
I made a few suggestions about documentation and naming, suggesting vertices instead of mesh to avoid confusion with `ignition::common::Mesh`

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh]: also here: I think we should check the size and exit early if there are no vertices

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (scpeters) [inline eigen3/include/ignition/math/eigen3/Util.hh:65]: this is uninitialized. I would consider returning `Eigen::Matrix3d::Zero()`

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/gsoc-2021-machine-learning-extension-to-ignition-gazebo/1070/1

**Comment by adlarkin** [comment]
> I'm trying to do a few verifications for testing

I went ahead and ran the [example set of points in the unit test](https://github.com/ignitionrobotics/ign-math/blob/377623315f5396bf9f51183e84fa31699e259185/eigen3/src/Util_TEST.cc#L28-L61) through [Open3D's `CreateFromPoints` method](https://github.com/isl-org/Open3D/blob/76c2baf9debd460900f056a9b51e9a80de9c0e64/cpp/open3d/geometry/BoundingVolume.h#L123-L129) for comparison. Here are the results I got (link to the code can be found [here](https://gist.github.com/adlarkin/b02bdfa9fd52ea63acb94ee38f164fb8)):

```
Box center is: 
0.423213
0.507995
 1.83459

Box extent (size) is: 
6.53071
4.12782
3.25527

roll: -2.01634
pitch: 0.0058244
yaw: 1.1554
```

It looks like the box's center and size are similar enough between `Open3D` and this PR, but the rotation seems like it could be an issue. Of course, this is assuming that `Open3D` should be considered a "source of truth", but I just wanted to mention it here in case this deserves further investigation. The code I linked shows how I compute roll, pitch and yaw from the rotation matrix that is provided by `Open3D` (hopefully my calculations are correct). cc @scpeters

**Comment by AmrElsersy** [comment]
That is our output for multi-links models (double pendolum and cars) 

image

**Comment by AmrElsersy** [comment]
> @adlarkin
> I thinks that is because the part of calculating the orientation in the 2 algorithms is different
> 
> their code
> 
> ```c++
>     Eigen::Matrix3d R = es.eigenvectors();
>     R.col(0) /= R.col(0).norm();
>     R.col(1) /= R.col(1).norm();
>     R.col(2) /= R.col(2).norm();
> 
>     if (evals(1) > evals(0)) {
>         std::swap(evals(1), evals(0));
>         Eigen::Vector3d tmp = R.col(1);
>         R.col(1) = R.col(0);
>         R.col(0) = tmp;
>     }
>     if (evals(2) > evals(0)) {
>         std::swap(evals(2), evals(0));
>         Eigen::Vector3d tmp = R.col(2);
>         R.col(2) = R.col(0);
>         R.col(0) = tmp;
>     }
>     if (evals(2) > evals(1)) {
>         std::swap(evals(2), evals(1));
>         Eigen::Vector3d tmp = R.col(2);
>         R.col(2) = R.col(1);
>         R.col(1) = tmp;
>     }
> ```

That was the output when I tried that code of Open3D for calculating Rotation instead of the current algorithm
It was bad

ezgif com-gif-maker (2)

File: eigen3/src/Util_TEST.cc
Comment (AmrElsersy) [reply eigen3/src/Util_TEST.cc:26]: mmmm the Simple box approach worked ! and gave me the exact box 

I tried that before but on a rotated box in all directions, and it didn't gave me the same box, so that is why I told you that it will not work

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (AmrElsersy) [reply eigen3/include/ignition/math/eigen3/Util.hh:44]: Ok

File: eigen3/include/ignition/math/eigen3/Util.hh
Comment (adlarkin) [inline eigen3/include/ignition/math/eigen3/Util.hh:129]: Is this one of the `math::eigen3::convert` methods? If so, we can add the `math::eigen3` namespace so that it's easy for code readers to know which function is being used (adding this namespace would also help avoid potential collisions since `convert` is a generic method name).

File: eigen3/src/Util_TEST.cc
Comment (adlarkin) [inline eigen3/src/Util_TEST.cc:87]:
```suggestion
  EXPECT_DOUBLE_EQ(rotation.Pitch(), 0);
  EXPECT_DOUBLE_EQ(rotation.Yaw(), 0);
```
nit: fix whitespace

File: eigen3/src/Util_TEST.cc
Comment (adlarkin) [inline eigen3/src/Util_TEST.cc:127]: Can you add one more test for `math::eigen3::covarianceMatrix` where an empty set of vertices is passed in? We should receive an identity matrix in this case.

**Review by adlarkin — COMMENTED** [review]
Just a few small comments, but I think this is almost ready to merge!

**Comment by scpeters** [comment]
have we considered specifying the orientation of the bounding box and then simply expanding its dimensions to include all the passed vertices? this would avoid the weird box rotations at the cost of generating non-minimal boxes

**Comment by adlarkin** [comment]
> have we considered specifying the orientation of the bounding box and then simply expanding its dimensions to include all the passed vertices? this would avoid the weird box rotations at the cost of generating non-minimal boxes

I think the challenge is that in some scenarios, we have the vertices of an object, but not the orientation of the object. For example, in https://github.com/ignitionrobotics/ign-rendering/pull/334, I believe that we have all of the vertices from the bounding boxes provided by Ogre, but I don't believe that we have a way of getting the "combined orientation" from all of the rendering visuals (I could be wrong about this).

However, that being said, I think that it would be good to provide users with the option to pass in an orientation to `verticesToOrientedBox`. Perhaps we can add a input parameter that's a pointer to an orientation: if the pointer is null, we ignore it and run the algorithm as-is, but if it isn't null, then we take the orientation into account. Does that sound reasonable?

**Review by adlarkin — COMMENTED** [review]
Code looks good! I'm going to wait to see what @scpeters thinks about https://github.com/ignitionrobotics/ign-math/pull/224#issuecomment-909325297 before merging this.

**Comment by scpeters** [comment]
> However, that being said, I think that it would be good to provide users with the option to pass in an orientation to `verticesToOrientedBox`. Perhaps we can add a input parameter that's a pointer to an orientation: if the pointer is null, we ignore it and run the algorithm as-is, but if it isn't null, then we take the orientation into account. Does that sound reasonable?

I would just add a separate API that accepts a vector of vertices as well as a `Quaternion`, but this PR is fine without it. We could add it later if desired

**Review by adlarkin — APPROVED** [review]

**Pull request merged**
